golang/cosmos/x/lien/lien.go

```diff
@@ -60,12 +77,77 @@ func (ch portHandler) Receive(ctx *vm.ControllerContext, str string) (string, er
 	case LIEN_GET_ACCOUNT_STATE:
 		return ch.handleGetAccountState(ctx.Context, msg)
 
+	case LIEN_GET_STAKING:
+		return ch.handleGetStaking(ctx.Context, msg)
+
 	case LIEN_SET_LIENED:
 		return ch.handleSetLiened(ctx.Context, msg)
 	}
 	return "", fmt.Errorf("unrecognized type %s", msg.Type)
 }
 
+func (ch portHandler) handleGetStaking(ctx sdk.Context, msg portMessage) (string, error) {
+	reply := msgStaking{
+		EpochTag:        fmt.Sprint(ctx.BlockHeight()),
+		Denom:           ch.keeper.BondDenom(ctx),
+		ValidatorValues: make([]sdk.Int, 0),
+		DelegatorStates: make([]delegatorState, 0),
+	}
+	validatorIndex := map[string]int{}
+	for i, v := range msg.Validators {
+		validatorIndex[v] = i
+		var value sdk.Int
+		valAddr, err := sdk.ValAddressFromBech32(v)
+		if err != nil {
+			value = sdk.NewInt(-1)
+		} else {
+			value = sdk.NewInt(0)
+			validator, found := ch.keeper.GetValidator(ctx, valAddr)
+			if found {
+				value = validator.Tokens
+			}
+		}
+		reply.ValidatorValues = append(reply.ValidatorValues, value)
+	}
+	for _, d := range msg.Delegators {
+		state := delegatorState{
+			ValidatorIdx: make([]int, 0),
+			Values:       make([]sdk.Int, 0),
```

Comment:
Again, preallocate for the worst case:
```suggestion
			ValidatorIdx: make([]int, 0, len(delegatorDelegations)),
			Values:       make([]sdk.Int, 0, len(delegatorDelegations)),
```

Reply:
Done - on the expectation that validators are likely to vote in the common case.